src/java.base/share/classes/java/util/zip/ZipFile.java

```diff
@@ -1402,14 +1404,44 @@ private static class Key {
 
             public Key(File file, BasicFileAttributes attrs, ZipCoder zc) {
                 this.attrs = attrs;
-                this.file = file;
+                Object fk = attrs.fileKey();
+                if (fk != null) {
+                    // use file value, hashCode will use fileKey()
+                    this.file = file;
+                } else {
+                    this.file = getCanonicalFile(file);
+                }
                 this.utf8 = zc.isUTF8();
             }
 
+            @SuppressWarnings("removal")
+            private File getCanonicalFile(File file) {
+                if (System.getSecurityManager() != null) {
+                    PrivilegedAction<File> pa = () -> {
+                        try {
+                            return file.getCanonicalFile();
+                        } catch (IOException e) {
+                            throw new RuntimeException(e);
```

Comment:
Mapping to RuntimeException doesn't look right here, instead need to use PrivilegedExceptionAction here and changing the Key constructor to throw IOException.